### ISSUE TITLE
Enable forced sampling of traces for a specific tenant (Http adapter)

### ISSUE THREAD

**Issue by calohmn** [eclipse/hono#896]
The OpenTracing Specification allows setting a specific tag (`sampling.priority` - see [Span tags](https://github.com/opentracing/specification/blob/master/semantic_conventions.md#span-tags-table)), 
with which it can be enforced that trace spans are actually captured in the Tracing component.

To enable debugging of problems related to messages of a specific tenant in Hono, it would be good to be able to define such an enforced trace sampling for all spans regarding that specific tenant.

For that, the tenant JSON structure, as defined in the Tenant API, could be extended with an additional, optional attribute: `tracing-sampling-priority` (integer). 
It's semantics should follow the one given for the `sampling.priority` tag:
````
If greater than 0, a hint to the Tracer to do its best to capture the trace. 
If 0, a hint to the trace to not-capture the trace. 
If absent, the Tracer should use its default sampling mechanism.
````
For completeness, I would suggest having a value of `-1` be equivalent with the value being absent.

This `tracing-sampling-priority` setting of a tenant should then be applied (in the sense of setting the `sampling.priority` tag) to the traces associated with that tenant.

EDIT: this issue shall track the implementation of this feature for the HTTP adapter only.

**Comment by calohmn** [comment]
I've done some testing with setting the `sampling.priority` tag.
It looks like Jaeger currently requires that tag to be present in all spans of a trace for the complete trace to be forcefully sampled. I.e. it is not enough to set the tag on one span (root span or a child span) and then have all the spans of that trace be automatically forcefully sampled.

That means that the information, whether spans should be forcefully sampled, has to be given to all methods where new child spans get created.
In Hono, there is currently the `SpanContext` object used as a method parameter in these methods.
As a first step (and first PR), I would suggest replacing these parameters with a new Object that holds `SpanContext` and the sampling priority.

(Thinking further about tracing across components, this all probably means that we should consider transferring the sampling priority information in a baggage item so that other components can implement the tenant-specific tracing priority as well.)

**Comment by calohmn** [comment]
In case there are a lot of traces being generated for a particular tenant, it would be a good feature to have an option to only force sampling of *error traces* for that particular tenant.

This could be implemented in these ways:
- A: Either with these attributes in the tenant json:
````
tracing-sampling-priority             [0|1|-1]    => optional, default -1 (i.e. use overall sampling strategy)
tracing-sampling-priority-applies-to  [error|all] => optional, default value could be "error"
````
- B: or with these:
````
tracing-sampling-priority-all     [0|1|-1]    => optional, default -1 (i.e. use overall sampling strategy)
 => sets the prio for all traces of the tenant
tracing-sampling-priority-error   [0|1|-1]    => optional, default -1 (i.e. use "-all" prio or overall sampling strategy)
 => sets the prio for error traces of the tenant, overriding the -all value
````

Option A would maybe look somewhat easier to the user.
Option B would be more flexible and would allow these additional configurations:
- sample *only* error traces for the tenant, disable sampling of non-error traces (-all=0,-error=1)
- sample all tenant traces, *except* error traces (-all=1,-error=0)

I would opt for Option B.

**Comment by sophokles73** [comment]
I would like to keep the technical details of the tracing framework out of the tenant configuration, if possible.

A third option comes to mind which might be easier to understand for users:
`trace-level                   [default|error-only|all] `

**Comment by calohmn** [comment]
Ok, right, having such names as values makes things clearer than having to use the integer values.

I only would rename the setting and add `none` and `default_all_errors` as possible values (and use `errors-only`):
`trace-sampling [default|default_all_errors|errors-only|all|none]`.
The `level` term isn't used in the OpenTracing context as far as I know, and might be confusing here.

**Comment by sophokles73** [comment]
I think with `error-only` I meant what you would call `default-all-errors` (which seems to be clearer). Based on that I do not think that we need `errors-only` and `none`.

**Comment by calohmn** [comment]
Ok. I would suggest to start out with `trace-sampling-mode [default|all]`. Adding `default-all-errors` (and possibly `errors-only` and `none`) could be done later. 

More interesting is probably adding the possibility to force trace-sampling just for a specific *device* as well. This information should be held in the device registry's device information, imho.
That means that `trace-sampling-mode` can either be set for a tenant, or for just one device.

Concerning the implementation, it looks like we should keep a cache for this information.
The reason is, that there are already spans being created *before* the `trace-sampling-mode` information from the `TenantService` and the `RegistrationService` is retrieved. These spans would not get sampled.

**Comment by calohmn** [comment]
Closed via #1382.
